### PR TITLE
Support multiple tokens locally

### LINKED ISSUE #2446: Support multiple tokens locally
_Originally from @osanseviero on [slack](https://huggingface.slack.com/archives/C02V5EA0A95/p1721119617996949) (private)_

> With the big push for fine-grained tokens, I'm wondering if we should have a way for users to have multiple tokens easily in their environment and switch between them rather than overriding each time

_Originally from @julien-c in reply_

> aws-cli has something like this

> we could add a flag for --profile token_name  or --token-name "My token to list private models"

_Originally from @osanseviero as well_

> On my side I stopped using Colab secret management entirely as it doesn't make sense in the fine-grained tokens world. I'm again signing in each time with a new key

---

We would need:
- to be able to define a name per token in `login`
- a new command / helper to list existing local tokens
- a new command / helper to switch the "current" local token

Google Colab secret management is a separate topic but still relevant.

### PR DESCRIPTION
**Pull request by hanouticelina**
Closes #2446.

This PR adds a multi-token support allowing users to easily manage and switch between multiple tokens.

## List of changes
- Added a `token-name` argument  `logout` helpers and the corresponding commands.
- Refactored all modules related to authentification into `huggingface_hub/src/utils/_auth.py`.
- Added new commands `huggingface-cli auth switch` and `huggingface-cli auth list` to switch between tokens and list saved tokens respectively. New helpers `huggingface_hub._login.auth_switch` and `huggingface_hub._login.auth_list` are also added.
- Multiple helper functions are added :  `get_stored_tokens()`, `_save_stored_tokens()`, `_get_token_by_name()`, `_save_token()`.
- Unit tests are added in `tests/test_auth.py`.
- Documentation: A new section added to `quick-start.md`, `guides/cli.md` and login package reference are also updated.

### Login
```bash
huggingface-cli login [--token TOKEN]
```
- Logs in with the provided token or prompts for one if not provided.
- The token is saved locally under the name provided by the server.
- If `--token` is not provided, prompts the user for token interactively.

### List locally stored tokens
```bash
huggingface-cli auth list
```
- Lists all saved tokens.

Example output:
```
name                     | token          
-------------------------|---------------
  dummy_token            | hf_****XXXX    
  * all_perm             | hf_****YYYY    


Note: Environment variable `HF_TOKEN` is set and is the current active token.
```
### Logout
```bash
huggingface-cli logout [--token-name TOKEN_NAME]
```
- Removes the specified token.
- If `--token-name` is not specified, logs out from all tokens.

### Switching the current active token
```bash
huggingface-cli auth switch [--token-name TOKEN_NAME]
```
- Sets the specified token as the active token for subsequent commands.
- If `--token-name` is not provided, prompts the user to select from locally saved tokens.
- This command will override the token written in `~/.cache/huggingface/token` (defined by `HF_TOKEN_PATH` env variable).

### Token storage
Tokens will be stored in `~/.cache/huggingface/stored_tokens` (or `HF_STORED_TOKENS_PATH` env variable if set), which is INI format file:
```ini
[token_name_1] 
hf_token = hf_XXXXXXX 
[token_name_2] 
hf_token = hf_YYYYYYY
[token_name_3] 
hf_token = hf_ZZZZZZZ
```

### Token Selection Priority
We will keep the same token retrieval priority order
1. `HF_TOKEN` env variable.
2. Token file `~/.cache/huggingface/token`. The token stored in this file will be updated when switching between tokens.

### CODE REVIEW

**Review by Wauplin — COMMENTED** [review]
Thanks for working on it @hanouticelina !

I've not properly tested it locally yet but started to make a pass on the implementation. I think we can regroup some things to avoid multiple utils modules and decouple the low-level logic (e.g. how the config file is structured) from the high-level logic (e.g. how to present stuff to the users).

File: src/huggingface_hub/_login.py [hunk 1]
```diff
@@ -49,6 +56,7 @@
 
 def login(
     token: Optional[str] = None,
+    profile_name: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (hanouticelina) [reply src/huggingface_hub/_login.py]: Yes, you're right! I'll definitely keep backward compatibility constraint more in mind in future changes

File: src/huggingface_hub/utils/_auth_profiles.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+from configparser import ConfigParser
+from pathlib import Path
+
+from .. import constants
+
+
+def _read_profiles() -> ConfigParser:
```
[the remark this answers is not in the document]
Comment (hanouticelina) [reply src/huggingface_hub/utils/_auth_profiles.py]: Indeed!

File: src/huggingface_hub/commands/user.py [hunk 3]
```diff
@@ -95,12 +139,51 @@ def __init__(self, args):
...
     def run(self):
-        logout()
+        logout(profile_name=self.args.profile_name)
+
+
+class AuthSwitchCommand(BaseUserCommand):
+    def run(self):
+        profile_name = self.args.profile_name
+        if profile_name is None and not self.args.disable_tui:
+            profile_name = self._select_profile_tui()
```
Comment (Wauplin) [inline src/huggingface_hub/commands/user.py:158]: Sorry for suggesting the TUI option. Maybe for a first PR that was not the best idea as it can lead to more friction :see_no_evil: What I would suggest here to avoid friction:
- remove `--disable-tui` option
- if InquirerPy installed => `self._select_profile_tui()`
- if not installed => a simple "print a numbered list" + "type a number" UI to select it

Therefore, `InquirerPy` is not a requirement anymore and not having it still leads to a good-enough interface.

File: src/huggingface_hub/commands/delete_cache.py [hunk 4]
```diff
@@ -77,24 +76,6 @@
     _inquirer_py_available = False
 
 
-def require_inquirer_py(fn: Callable) -> Callable:
```
Comment (Wauplin) [inline src/huggingface_hub/commands/delete_cache.py:80]: (changes not needed in the end if we make inquire_py optional in auth switch, sorry about that :see_no_evil: )

File: src/huggingface_hub/utils/_auth.py [hunk 5]
```diff
@@ -121,6 +125,88 @@ def _get_token_from_file() -> Optional[str]:
...
+        profiles (`Dict[str, str]`):
+            The profiles to save. Key is the profile name and value is the token.
+    """
+    profiles_path = Path(constants.HF_PROFILES_PATH)
+
+    # Write the profiles into an INI file
+    config = configparser.ConfigParser()
+    for profile_name, token in profiles.items():
+        config.add_section(profile_name)
+        config.set(profile_name, "hf_token", token)
```
Comment (Wauplin) [inline src/huggingface_hub/utils/_auth.py:167]:
```suggestion
    for profile in sorted(profiles.keys()):
        config.add_section(profile)
        config.set(profile, "hf_token", profiles[profile])
```

(nit) for better UX, I would sort the profiles in the file. This way they'll be displayed sorted in the UI as well. Usually makes it easier to find the correct one

File: tests/test_auth.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
+import os
```
Comment (Wauplin) [inline tests/test_auth.py:1]: Really nice test suite! I few comments I have:
- no need for the classes. In my opinion, it makes things less readable. I know we have plenty of them in other modules but let's consider them as legacy :D When outside of a test class you can use fixtures more easily
- I like the idea of not using the "real" token and profiles paths. To go one step further, we could use a "per-test autouse fixture" that creates a path in tmp directory for each test. The problem I see with the current is if the tests are ran in parallel (which is the case in the CI), they might conflicts with each other

Apart from that, thanks for the complete tests!

**Review by Wauplin — COMMENTED** [review]
Hey there! I finally took some time to extensively test things locally. I think you are on the right track to get this feature shipped. I found a few corner cases that I think should be tackled differently. Other than that, it's mostly cosmetic changes in the API design

**Comment by Wauplin** [comment]
(btw, you can set the PR to "ready for review" when it's the case :) )

File: src/huggingface_hub/_login.py [hunk 7]
```diff
@@ -108,31 +123,47 @@ def login(
...
+            profile_name=profile_name,
+        )
     elif is_notebook():
         notebook_login(new_session=new_session, write_permission=write_permission)
     else:
         interpreter_login(new_session=new_session, write_permission=write_permission)
 
 
-def logout() -> None:
+def logout(profile_name: Optional[str] = None) -> None:
```
[the remark this answers is not in the document]
Comment (hanouticelina) [reply src/huggingface_hub/_login.py:138]: I found it also a little bit more annoying in the CLI, I hesitated but ended up keeping profile_name for clarity

File: src/huggingface_hub/_login.py [hunk 8]
```diff
@@ -312,22 +414,73 @@ def _login(token: str, add_to_git_credential: bool, write_permission: bool = Fal
...
+    profiles = get_profiles()
+    # If there is no profiles saved or the profile name is not found, do nothing
+    if not profiles or profile_name not in profiles:
+        return
+
+    token = profiles[profile_name]
+    del profiles[profile_name]
+    _save_profiles(profiles)
+
+    if token == get_token():
```
[the remark this answers is not in the document]
Comment (hanouticelina) [reply src/huggingface_hub/_login.py:450]: indeed!

File: tests/test_auth.py (hunk 6 above)
Comment (Wauplin) [reply tests/test_auth.py:1]: My bad, I was thinking of unittest classes, no just raw classes.
Using basic classes to group things is fine yes (not so opinionated to be honest ^^). The problem is when you inherit from `unittest.TestCase` (which is what we do everywhere). When doing so, the fixtures cannot be used:

```py
import unittest

import pytest


@pytest.fixture
def hello():
    return "Hello World"

def test_something(hello: str):
    assert hello == "Hello World"

class TestSomethingInClass():
    def test_something(self, hello: str):
        assert hello == "Hello World"

class TestSomethingInUnitTestClass(unittest.TestCase):
    def test_something(self, hello: str):
        assert hello == "Hello World"
```

will lead to 

```
(...)

self = <test.TestSomethingInUnitTestClass testMethod=test_something>
method = <bound method TestSomethingInUnitTestClass.test_something of <test.TestSomethingInUnitTestClass testMethod=test_something>>

    def _callTestMethod(self, method):
>       method()
E       TypeError: TestSomethingInUnitTestClass.test_something() missing 1 required positional argument: 'hello'

/usr/lib/python3.10/unittest/case.py:549: TypeError
================================================================================ slowest durations ================================================================================

(9 durations < 0.005s hidden.  Use -vv to show these durations.)
============================================================================= short test summary info =============================================================================
FAILED test.py::TestSomethingInUnitTestClass::test_something - TypeError: TestSomethingInUnitTestClass.test_something() missing 1 required positional argument: 'hello'
=========================================================================== 1 failed, 2 passed in 0.26s ===========================================================================
```

Comment (Wauplin) [reply tests/test_auth.py:1]: There are ways to mitigate unittest classes with pytest fixture but not very satisfying (see https://docs.pytest.org/en/latest/how-to/unittest.html#using-autouse-fixtures-and-accessing-other-fixtures). It works only for autouse and applies to all tests at once.

**Review by Wauplin — COMMENTED** [review]
Looking good! As discussed offline, let's use the "token name" returned by the server as profile names :) Once updated we should be good to merge. Thanks for the extensive tests!

**Review by Wauplin — APPROVED** [review]
Thanks for the change! Last round of review I think :) I tested everything locally and the UX is really smooth now :hugs: 

Feel free to merge once the comments below are addressed and CI is green.

File: tests/test_auth.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+class TestSetActiveToken:
+    def test_set_active_token_success(self):
+        _save_token(TOKEN, "test_token")
+        _set_active_token("test_token", add_to_git_credential=False)
+        assert _get_token_from_file() == TOKEN
+
+    def test_set_active_token_non_existent(self):
+        non_existent_token = "non_existent"
+        with pytest.raises(
+            ValueError, match=rf"Token {non_existent_token} not found in {constants.HF_STORED_TOKENS_PATH}"
```
Comment (Wauplin) [inline tests/test_auth.py:74]:
```suggestion
            ValueError, match="Token non_existent not found in .*"
```

this should fix the CI failure on Windows and is more than enough to test the ValueError in case of non existent

File: src/huggingface_hub/_login.py [hunk 10]
```diff
@@ -312,6 +404,57 @@ def _login(token: str, add_to_git_credential: bool, write_permission: bool = Fal
...
+        [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError):
+            If the access token name is not found.
+    """
+    stored_tokens = get_stored_tokens()
+    # If there is no access tokens saved or the access token name is not found, do nothing
+    if not stored_tokens or token_name not in stored_tokens:
+        return
+
+    token = stored_tokens[token_name]
+    del stored_tokens[token_name]
```
[the remark this answers is not in the document]
Comment (hanouticelina) [reply src/huggingface_hub/_login.py:437]: yep, tend to use `del dict[key]` instead of `value = dict.pop(key)` when I know that the key exists since it's a little bit faster than dict.pop()

**Comment by hanouticelina** [comment]
thanks a lot @Wauplin and sorry for the lengthy back and forth reviews 🙈 with your detailed feedback on this PR, I'm sure the next ones will be much easier to review 😄 
failing tests seem unrelated

**Review by Wauplin — APPROVED** [review]
All good!

**Comment by hanouticelina** [comment]
I think we are good with this one, let's merge :)

**Comment by Pierrci** [comment]
nice!! 🎉🎉🎉
